### PR TITLE
[pip][design] PIP-278: Support pluggable topic compaction service

### PR DESCRIPTION
**Pull request by coderzc**
### Motivation & Modifications

Start a PIP: Support pluggable topic compaction service

### Documentation
- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete`

### CODE REVIEW

File: pip/pip-278.md
Comment (mattisonchao) [inline pip/pip-278.md:33]: Would you mind putting an Inheritance chart to help others understand this part? :)

File: pip/pip-278.md
Comment (mattisonchao) [inline pip/pip-278.md:47]: Should we consider defining some exceptions or status codes here?

File: pip/pip-278.md
Comment (coderzc) [reply pip/pip-278.md:35]: Yeah, I will add a class diagram

File: pip/pip-278.md
Comment (mattisonchao) [inline pip/pip-278.md:71]: It looks like ` CompletableFuture<Entry> readCompactedLastEntry();` includes `CompletableFuture<PositionImpl> getCompactedLastPosition();`? 

The compacted last entry includes its position. Isn't it? :)

File: pip/pip-278.md
Comment (mattisonchao) [inline pip/pip-278.md:64]: Should we consider adding `@NonNull` annotation to interface parameters to clarify it? :)

File: pip/pip-278.md
Comment (mattisonchao) [inline pip/pip-278.md:99]: Hi，@asafm 
I remember you have been working on the metrics lately. What do you think of the place?  thanks! :)

File: pip/pip-278.md
Comment (coderzc) [reply pip/pip-278.md:71]: Yes, but sometimes the cost of `getCompactedLastPosition` may be small or can get it synchronous.
Separating them is for performance.

File: pip/pip-278.md
Comment (coderzc) [reply pip/pip-278.md:64]: Add `@NonNull` annotation for `List<Entry>`? It seems to be unable to add an annotation for the Generic Types

Comment (mattisonchao) [reply pip/pip-278.md:66]: add annotation to `PositionImpl startPosition`

File: pip/pip-278.md
Comment (mattisonchao) [reply pip/pip-278.md:73]: Got it.

File: pip/pip-278.md
Comment (coderzc) [reply pip/pip-278.md:66]: Oh, Good idea

**Review by BewareMyPower — COMMENTED** [review]
It's better not to use an implementation class in the interface. Why do you use `PositionImpl` instead of `Position`?

File: pip/pip-278.md
Comment (coderzc) [reply pip/pip-278.md:49]: It doesn't seem to be necessary, we don't have any special handling of exceptions for these methods

**Comment by coderzc** [comment]
> It's better not to use an implementation class in the interface. Why do you use `PositionImpl` instead of `Position`?

Oh, Because when implementing it, I found that the caller needs to compare the `Position`.
Please see: https://github.com/coderzc/pulsar/blob/feat_compactedService/pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java#L1386

**Comment by coderzc** [comment]
I push PR for implementation: https://github.com/apache/pulsar/pull/20626

**Comment by BewareMyPower** [comment]
> the caller needs to compare the Position.

It cannot explain why must we return a `PositionImpl` instead of `Position`.

**Comment by coderzc** [comment]
> > the caller needs to compare the Position.
> 
> It cannot explain why must we return a `PositionImpl` instead of `Position`.

Just `PositionImpl` implemented the `Comparable` interface, if return `Position` we need to strong transform  to `PositionImpl`

File: pip/pip-278.md
Comment (codelipenghui) [inline pip/pip-278.md:3]: It should be a wrong description of the topic compaction. It is not for removing the duplicated messages. The message deduplication was. The topic compaction provides a key-based data retention mechanism that allows you only to keep the most recent message associated with that key.

Comment (codelipenghui) [reply pip/pip-278.md:3]: And we should also add the topic compaction use case of the new load balancer. It has improved the topic compaction only to keep the first value of the key.

And we should also explain if the new compaction service will affect the load balancer or not.

File: pip/pip-278.md
Comment (codelipenghui) [inline pip/pip-278.md:10]: We should explain why Pulsar topic compaction can't work with KoP. Actually, it can work with pulsar format data in KoP, but it can't work with Kafka format data since the data written to the entry is in Kafka format. The Pulsar compactor doesn't aware of Kafka format data. And it doesn't make sense to support Kafka format data handling in Pulsar.

File: pip/pip-278.md
Comment (codelipenghui) [inline pip/pip-278.md:12]: I think this should be a further consideration. We don't have any discussion about Pulsar NEED to support S3, columnar format, or partitioning.

File: pip/pip-278.md
Comment (codelipenghui) [inline pip/pip-278.md:73]:
```suggestion
      CompletableFuture<Entry> readLastCompactedEntry();
```

File: pip/pip-278.md
Comment (codelipenghui) [inline pip/pip-278.md:80]:
```suggestion
      CompletableFuture<Position> getLastCompactedPosition();
```

File: pip/pip-278.md
Comment (codelipenghui) [inline pip/pip-278.md:124]: I think we have alternatives.

- Only make the compactor pluggable
- Make the compaction data serializer and deserializer pluggable

But they are not good for the long-term view of Pulsar

Comment (coderzc) [reply pip/pip-278.md:124]: Yeah, I add this part

File: pip/pip-278.md
Comment (coderzc) [reply pip/pip-278.md:10]: Yesh, I add some description to explain why Pulsar topic compaction can't work with KoP.

File: pip/pip-278.md
Comment (coderzc) [reply pip/pip-278.md:3]: Ok, I fix description of the topic compaction

File: pip/pip-278.md
Comment (codelipenghui) [inline pip/pip-278.md:6]: Please help add a link of the proposal of new load balancer so that users can go to the proposal to check the details.

Comment (codelipenghui) [inline pip/pip-278.md:6]:
```suggestion
Another Pulsar's internal use case, the Topic Compaction of new load balancer changed the strategy of compaction, it will only to keep the first value of the key.
```

**Review by codelipenghui — COMMENTED** [review]
Cool, just left a minor comment. 
It looks good to me to go to the VOTE stage.

**Review by codelipenghui — COMMENTED** [review]
The Class Diagram should be updated since you have changed the method name

**Comment by coderzc** [comment]
> The Class Diagram should be updated since you have changed the method name

Thanks, I fixed it.

**Review by codelipenghui — APPROVED** [review]

**Review by mattisonchao — APPROVED** [review]
LGTM

**Comment by mattisonchao** [comment]
@Anonymitaet please help take a look at this proposal. thanks a lot!
 :)

File: pip/pip-278.md
Comment (BewareMyPower) [inline pip/pip-278.md:71]:
```suggestion
      CompletableFuture<List<Entry>> readCompactedEntries(Position startPosition, int numberOfEntriesToRead);
```

We should not add a `NonNull` annotation on a interface. If the annotation comes from lombok, it means a null check will be performed, see https://projectlombok.org/features/NonNull.

Instead, we should document the exact behavior when `startPosition` is `null` via a `@throws` or `@implNote` annotation.

File: pip/pip-278.md
Comment (BewareMyPower) [inline pip/pip-278.md:68]:
```suggestion
       * @param numberOfEntriesToRead the maximum number of entries to read.
```

I guess you want to mean the "maximum" number of entries. If not, you should explain the behavior that what if there are not enough entries to read.

File: pip/pip-278.md
Comment (BewareMyPower) [inline pip/pip-278.md:76]:
```suggestion
       * @return a future that will be completed with the compacted last entry, this entry can be null.
```

Correct the grammar.

File: pip/pip-278.md
Comment (BewareMyPower) [inline pip/pip-278.md:83]:
```suggestion
       * @return a future that will be completed with the last compacted position, this position can be null.
```

File: pip/pip-278.md
Comment (BewareMyPower) [inline pip/pip-278.md:85]: It's confusing that we have both `readLastCompactedEntry` and `getLastCompactedPosition` because we have a `Entry#getPosition` method to get the position of an entry.

Could you explain the difference between:

```java
readLastCompactedEntry().thenApply(Entry::getPosition)
```

and 

```java
getLastCompactedPosition()
```

Comment (coderzc) [reply pip/pip-278.md:85]: Please ref to: https://github.com/apache/pulsar/pull/20624#discussion_r1236177136

File: pip/pip-278.md
Comment (coderzc) [reply pip/pip-278.md:71]: > We should not add a NonNull annotation on a interface. If the annotation comes from lombok, it means a null check will be performed, see https://projectlombok.org/features/NonNull.

`Nonnull` means [`javax.annotation.Nonnull`](https://docs.oracle.com/javaee/7/api/javax/validation/constraints/NotNull.html), it only documents this element must not be null

File: pip/pip-278.md
Comment (coderzc) [reply pip/pip-278.md:68]: Yeah, mean the "maximum" number of entries.

Comment (coderzc) [reply pip/pip-278.md]: Fixed.

**Review by BewareMyPower — APPROVED** [review]

**Comment by asafm** [comment]
@BewareMyPower Why approve if the vote hasn't been concluded?

**Comment by BewareMyPower** [comment]
@asafm I see @codelipenghui and @mattisonchao approved before so I think it's equivalent with a +1 and I also voted a binding +1 in the ML after I approved. Before asking me, I think you should ask them first.

From https://github.com/apache/pulsar/tree/master/pip, you can see

> When the vote is closed, if the outcome is positive, ask a PMC member (using voting thread on mailing list) to merge the PR.

I think it's okay to approve the PIP PR before the vote is closed, but we should only merge the PIP PR after the vote is closed.

**Comment by asafm** [comment]
@BewareMyPower Ok.

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:6]:
```suggestion
Another Pulsar's internal use case, the Topic Compaction of the new load balancer, changed the strategy of compaction. It only keeps the first value of the key. For more detail, see [PIP-215](https://github.com/apache/pulsar/issues/18099).
```

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:12]:
```suggestion
Currently, the implementation of Pulsar Topic Compaction is fixed and does not support customized strategy, which limits users from using more Compactor policies in their applications.
```
Do you mean this?

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:15]:
```suggestion
The Pulsar compactor doesn't aware of Kafka format data. And it doesn't make sense to support Kafka format data handling in Pulsar. We need to implement a pluggable compactor to support Kafka format data handling in KoP.
```

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:17]:
```suggestion
Another long-term consideration is that we may need to support writing the compacted data anywhere, S3, in columnar format, or even partitioning.
```

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:19]:
```suggestion
So we need to make the whole topic compaction service (including Write API & Read API) pluggable to support more customize compaction service implementation.
```

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:33]:
```suggestion
* For CompactorMetrics, keep the current implementation and don't define related methods in the topic compaction service interface. In the future, it will use the `Otel` interface or other metrics API instead.
```

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:40]:
```suggestion
To make the whole topic compaction service pluggable, we need to abstract `TopicCompactionService` interface. It ensures that the compactor provided a read API to read entries from compacted data.
```

Is this way more concise?

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:42]:
```suggestion
We should combine `CompactedTopicImpl` and `TwoPhaseCompactor` to the Pulsar implementation of the topic compaction service and make behavior with the current implementation consistent.
```

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:59]:
```suggestion
       * Topic Compaction is a key-based retention mechanism. It keeps the most recent value for a given key and
```

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:60]:
```suggestion
       * user reads compacted data from TopicCompactionService.
```

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:122]:
```suggestion
* For `CompactorMetrics`: keep the current implementation. Currently, it only supports `PulsarTopicCompactionService`. In the future, it will use the `Otel` API or other metrics API instead, and customize TopicCompactedService should implement it.
```
1. Do you mean this?

2. What do you intend to convey? "and customize TopicCompactedService should implement it."

File: pip/pip-278.md
Comment (Anonymitaet) [inline pip/pip-278.md:147]:
```suggestion
* Make the compaction data serializer and deserializer pluggable in the current Pulsar implementation.
```

File: pip/pip-278.md
Comment (coderzc) [reply pip/pip-278.md:40]: `TopicCompactionService` provides both the capability that the compactor and read entries from compacted data.

File: pip/pip-278.md
Comment (coderzc) [reply pip/pip-278.md:122]: > What do you intend to convey? "and customize TopicCompactedService should implement it."

Customized `TopicCompactedService` should implement the `Otel` API or other metrics API.

Comment (Anonymitaet) [reply pip/pip-278.md]: Then that sentence can be:
 "In the future, `CompactorMetrics` will use the `Otel` API or other metrics API. Customized `TopicCompactedService` should implement the Otel API or other metrics API."

It's better to show the objects explicitly rather than using a fuzzy word like `it`. Users do not what does "it" refer to.

Comment (coderzc) [reply pip/pip-278.md]: Thanks, I improve it.

**Review by Anonymitaet — APPROVED** [review]

**Comment by asafm** [comment]
Please bear in mind if for one PIP we have 0.5mb image, imaging the size over time in the repo.
It's forever there.

We could have used https://mermaid.js.org/syntax/classDiagram.html

**Comment by coderzc** [comment]
> Please bear in mind if for one PIP we have 0.5mb image, imaging the size over time in the repo. It's forever there.
> 
> We could have used https://mermaid.js.org/syntax/classDiagram.html

@asafm Thanks for reminding, I remove pip images and use mermaid instead: https://github.com/apache/pulsar/pull/20682

**Comment by asafm** [comment]
@coderzc https://github.com/apache/pulsar/pull/20684

**Pull request merged**
